File: php/phpsearch/src/phpsearch/SearchOptions.php
<?php

declare(strict_types=1);

namespace phpsearch;

use phpfind\ArgTokenizer;
use phpfind\ArgTokenType;
use phpfind\FileUtil;
use phpfind\FindException;

/**
 * Class SearchOptions
 */
class SearchOptions
{
    /**
     * @var SearchOption[] $options
     */
    private array $options;
    /**
     * @var array<string, \Closure> $bool_action_map
     */
    private readonly array $bool_action_map;
    /**
     * @var array<string, \Closure> $str_action_map
     */
    private readonly array $str_action_map;
    /**
     * @var array<string, \Closure> $int_action_map
     */
    private readonly array $int_action_map;
    /**
     * @var ArgTokenizer $arg_tokenizer
     */
    private ArgTokenizer $arg_tokenizer;


    /**
     * @throws SearchException
     */
    public function __construct()
    {
        $this->options = [];

        $this->bool_action_map = [
            'allmatches' => fn (bool $b, SearchSettings $ss) => $ss->first_match = !$b,
            'archivesonly' => fn (bool $b, SearchSettings $ss) => $ss->set_archives_only($b),
            'colorize' => fn (bool $b, SearchSettings $ss) => $ss->colorize = $b,
            'debug' => fn (bool $b, SearchSettings $ss) => $ss->set_debug($b),
            'excludehidden' => fn (bool $b, SearchSettings $ss) => $ss->include_hidden = !$b,
            'firstmatch' => fn (bool $b, SearchSettings $ss) => $ss->first_match = $b,
            'followsymlinks' => fn (bool $b, SearchSettings $ss) => $ss->follow_symlinks = $b,
            'help' => fn (bool $b, SearchSettings $ss) => $ss->print_usage = $b,
            'includehidden' => fn (bool $b, SearchSettings $ss) => $ss->include_hidden = $b,
            'multilinesearch' => fn (bool $b, SearchSettings $ss) => $ss->multi_line_search = $b,
            'nocolorize' => fn (bool $b, SearchSettings $ss) => $ss->colorize = !$b,
            'nofollowsymlinks' => fn (bool $b, SearchSettings $ss) => $ss->follow_symlinks = !$b,
            'noprintdirs' => fn (bool $b, SearchSettings $ss) => $ss->print_dirs = !$b,
            'noprintfiles' => fn (bool $b, SearchSettings $ss) => $ss->print_files = !$b,
            'noprintlines' => fn (bool $b, SearchSettings $ss) => $ss->print_lines = !$b,
            'noprintmatches' => fn (bool $b, SearchSettings $ss) => $ss->print_matches = !$b,
            'noprintresults' => fn (bool $b, SearchSettings $ss) => $ss->print_results = !$b,
            'norecursive' => fn (bool $b, SearchSettings $ss) => $ss->recursive = !$b,
            'nosearcharchives' => fn (bool $b, SearchSettings $ss) => $ss->set_search_archives(!$b),
            'printdirs' => fn (bool $b, SearchSettings $ss) => $ss->print_dirs = $b,
            'printfiles' => fn (bool $b, SearchSettings $ss) => $ss->print_files = $b,
            'printlines' => fn (bool $b, SearchSettings $ss) => $ss->print_lines = $b,
            'printmatches' => fn (bool $b, SearchSettings $ss) => $ss->print_matches = $b,
            'printresults' => fn (bool $b, SearchSettings $ss) => $ss->print_results = $b,
            'recursive' => fn (bool $b, SearchSettings $ss) => $ss->recursive = $b,
            'searcharchives' => fn (bool $b, SearchSettings $ss) => $ss->set_search_archives($b),
            'sort-ascending' => fn (bool $b, SearchSettings $ss) => $ss->sort_descending = !$b,
            'sort-caseinsensitive' => fn (bool $b, SearchSettings $ss) => $ss->sort_case_insensitive = $b,
            'sort-casesensitive' => fn (bool $b, SearchSettings $ss) => $ss->sort_case_insensitive = !$b,
            'sort-descending' => fn (bool $b, SearchSettings $ss) => $ss->sort_descending = $b,
            'uniquelines' => fn (bool $b, SearchSettings $ss) => $ss->unique_lines = $b,
            'verbose' => fn (bool $b, SearchSettings $ss) => $ss->verbose = $b,
            'version' => fn (bool $b, SearchSettings $ss) => $ss->print_version = $b
        ];

        $this->str_action_map = [
            'encoding' => fn (string $s, SearchSettings $ss) => $ss->text_file_encoding = $s,
            'in-archiveext' => fn (string $s, SearchSettings $ss) => $ss->add_exts($s, $ss->in_archive_extensions),
            'in-archivefilepattern' =>
                fn (string $s, SearchSettings $ss) => $ss->add_patterns($s, $ss->in_archive_file_patterns),
            'in-dirpattern' =>
                fn (string $s, SearchSettings $ss) => $ss->add_patterns($s, $ss->in_dir_patterns),
            'in-ext' => fn (string $s, SearchSettings $ss) => $ss->add_exts($s, $ss->in_extensions),
            'in-filepattern' =>
                fn (string $s, SearchSettings $ss) => $ss->add_patterns($s, $ss->in_file_patterns),
            'in-filetype' => fn (string $s, SearchSettings $ss) => $ss->add_file_types($s, $ss->in_file_types),
            'in-linesafterpattern' =>
                fn (string $s, SearchSettings $ss) => $ss->add_patterns($s, $ss->in_lines_after_patterns),
            'in-linesbeforepattern' =>
                fn (string $s, SearchSettings $ss) => $ss->add_patterns($s, $ss->in_lines_before_patterns),
            'linesaftertopattern' =>
                fn (string $s, SearchSettings $ss) => $ss->add_patterns($s, $ss->lines_after_to_patterns),
            'linesafteruntilpattern' =>
                fn (string $s, SearchSettings $ss) => $ss->add_patterns($s, $ss->lines_after_until_patterns),
            'maxlastmod' => fn (string $s, SearchSettings $ss) => $ss->max_last_mod = new \DateTime($s),
            'minlastmod' => fn (string $s, SearchSettings $ss) => $ss->min_last_mod = new \DateTime($s),
            'out-archiveext' =>
                fn (string $s, SearchSettings $ss) => $ss->add_exts($s, $ss->out_archive_extensions),
            'out-archivefilepattern' =>
                fn (string $s, SearchSettings $ss) => $ss->add_patterns($s, $ss->out_archive_file_patterns),
            'out-dirpattern' =>
                fn (string $s, SearchSettings $ss) => $ss->add_patterns($s, $ss->out_dir_patterns),
            'out-ext' => fn (string $s, SearchSettings $ss) => $ss->add_exts($s, $ss->out_extensions),
            'out-filepattern' =>
                fn (string $s, SearchSettings $ss) => $ss->add_patterns($s, $ss->out_file_patterns),
            'out-filetype' => fn (string $s, SearchSettings $ss) => $ss->add_file_types($s, $ss->out_file_types),
            'out-linesafterpattern' =>
                fn (string $s, SearchSettings $ss) => $ss->add_patterns($s, $ss->out_lines_after_patterns),
            'out-linesbeforepattern' =>
                fn (string $s, SearchSettings $ss) => $ss->add_patterns($s, $ss->out_lines_before_patterns),
            'path' => fn (string $s, SearchSettings $ss) => $ss->paths[] = $s,
            'searchpattern' =>
                fn (string $s, SearchSettings $ss) => $ss->add_patterns($s, $ss->search_patterns),
            'settings-file' => fn (string $s, SearchSettings $ss) => $this->update_settings_from_file($ss, $s),
            'sort-by' => fn (string $s, SearchSettings $ss) => $ss->set_sort_by($s)
        ];

        $this->int_action_map = [
            'linesafter' => fn (int $i, SearchSettings $ss) => $ss->lines_after = $i,
            'linesbefore' => fn (int $i, SearchSettings $ss) => $ss->lines_before = $i,
            'maxdepth' => fn (int $i, SearchSettings $ss) => $ss->max_depth = $i,
            'maxlinelength' => fn (int $i, SearchSettings $ss) => $ss->max_line_length = $i,
            'maxsize' => fn (int $i, SearchSettings $ss) => $ss->max_size = $i,
            'mindepth' => fn (int $i, SearchSettings $ss) => $ss->min_depth = $i,
            'minsize' => fn (int $i, SearchSettings $ss) => $ss->min_size = $i
        ];

        $this->set_options_from_json();
        $this->arg_tokenizer = new ArgTokenizer($this->options);
    }

    /**
     * @throws SearchException
     */
    private function set_options_from_json(): void
    {
        $search_options_path = FileUtil::expand_path(Config::SEARCH_OPTIONS_PATH);
        if (file_exists($search_options_path)) {
            $contents = file_get_contents($search_options_path);
            if ($contents === false || trim($contents) === '') {
                return;
            }

            try {
                $json_obj = json_decode(trim($contents), true, 512, JSON_THROW_ON_ERROR);
                $search_options = $json_obj['searchoptions'];
                if ($search_options) {
                    foreach ((array)$search_options as $so) {
                        $so = (array)$so;
                        $short = '';
                        if (array_key_exists('short', $so)) {
                            $short = (string)$so['short'];
                        }
                        $long = (string)$so['long'];
                        $desc = (string)$so['desc'];
                        if (array_key_exists($long, $this->bool_action_map)) {
                            $arg_type = ArgTokenType::Bool;
                        } elseif (array_key_exists($long, $this->str_action_map)) {
                            $arg_type = ArgTokenType::Str;
                        } elseif (array_key_exists($long, $this->int_action_map)) {
                            $arg_type = ArgTokenType::Int;
                        } else {
                            throw new SearchException('Invalid option: ' . $long);
                        }
                        $this->options[] = new SearchOption($short, $long, $desc, $arg_type);
                    }
                    usort($this->options, array('phpsearch\SearchOptions', 'cmp_search_options'));
                }

            } catch (\JsonException $e) {
                throw new SearchException($e->getMessage());
            }
        } else {
            throw new SearchException('File not found: ' . $search_options_path);
        }
    }

    /**
     * @param SearchSettings $settings
     * @param array $arg_tokens
     * @return void
     * @throws SearchException
     */
    private function update_settings_from_arg_tokens(SearchSettings $settings, array $arg_tokens): void
    {
        foreach ($arg_tokens as $arg_token) {
            if ($arg_token->type == ArgTokenType::Bool) {
                if (is_bool($arg_token->value)) {
                    $this->bool_action_map[$arg_token->name]($arg_token->value, $settings);
                    if (in_array($arg_token->name, array("help", "version"))) {
                        return;
                    }
                } else {
                    throw new SearchException("Invalid value for option: $arg_token->name");
                }
            } elseif ($arg_token->type == ArgTokenType::Str) {
                if (is_array($arg_token->value)) {
                    foreach ($arg_token->value as $val) {
                        if (is_string($val)) {
                            $this->str_action_map[$arg_token->name]($val, $settings);
                        } else {
                            throw new SearchException("Invalid value for option: $arg_token->name");
                        }
                    }
                } elseif (is_string($arg_token->value)) {
                    $this->str_action_map[$arg_token->name]($arg_token->value, $settings);
                } else {
                    throw new SearchException("Invalid value for option: $arg_token->name");
                }
            } elseif ($arg_token->type == ArgTokenType::Int) {
                if (is_int($arg_token->value)) {
                    $this->int_action_map[$arg_token->name]($arg_token->value, $settings);
                } else {
                    throw new SearchException("Invalid value for option: $arg_token->name");
                }
            } else {
                throw new SearchException("Invalid option: $arg_token->name");
            }
        }
    }

    /**
     * @param SearchSettings $settings
     * @param string $json
     * @return void
     * @throws SearchException
     */
    public function update_settings_from_json(SearchSettings $settings, string $json): void
    {
        try {
            $arg_tokens = $this->arg_tokenizer->tokenize_json($json);
            $this->update_settings_from_arg_tokens($settings, $arg_tokens);
        } catch (FindException $e) {
            throw new SearchException($e->getMessage());
        }
    }

    /**
     * @param SearchSettings $settings
     * @param string $file_path
     * @return void
     * @throws SearchException
     */
    private function update_settings_from_file(SearchSettings $settings, string $file_path): void
    {
        try {
            $arg_tokens = $this->arg_tokenizer->tokenize_file($file_path);
            $this->update_settings_from_arg_tokens($settings, $arg_tokens);
        } catch (FindException $e) {
            throw new SearchException($e->getMessage());
        }
    }

    /**
     * @param SearchSettings $settings
     * @param string[] $args
     * @return void
     * @throws SearchException
     */
    public function update_settings_from_args(SearchSettings $settings, array $args): void
    {
        try {
            $arg_tokens = $this->arg_tokenizer->tokenize_args($args);
            $this->update_settings_from_arg_tokens($settings, $arg_tokens);
        } catch (FindException $e) {
            throw new SearchException($e->getMessage());
        }
    }

    /**
     * @param string[] $args
     * @return SearchSettings
     * @throws SearchException
     */
    public function settings_from_args(array $args): SearchSettings
    {
        $settings = new SearchSettings();
        // default print_results to true since running as cli
        $settings->print_results = true;
        $this->update_settings_from_args($settings, $args);
        return $settings;
    }

    private static function cmp_search_options(SearchOption $o1, SearchOption $o2): int
    {
        return strcmp($o1->sort_arg, $o2->sort_arg);
    }

    /**
     * @return string
     */
    private function get_usage_string(): string
    {
        $usage = "Usage:\n phpsearch [options] -s <searchpattern>";
        $usage .= " <path> [<path> ...]\n\nOptions:\n";
        $opt_map = [];
        $longest = 0;
        foreach ($this->options as $option) {
            $opt_str = '';
            if ($option->short_arg) {
                $opt_str = '-' . $option->short_arg . ',';
            }
            $opt_str .= '--' . $option->long_arg;
            if (strlen($opt_str) > $longest) {
                $longest = strlen($opt_str);
            }
            $opt_map[$opt_str] = $option->desc;
        }
        $format_str = " %-" . $longest . "s  %s\n";
        foreach ($opt_map as $opt => $desc) {
            $usage .= sprintf($format_str, $opt, $desc);
        }
        return $usage;
    }

    /**
     * @return void
     */
    public function usage(): void
    {
        echo $this->get_usage_string() . "\n";
    }

    /**
     * @param int $exit_code
     * @return never
     */
    public function usage_and_exit(int $exit_code = 0): never
    {
        $this->usage();
        exit($exit_code);
    }
}

File: objc/objcsearch/Sources/objcsearch/SearchOptions.m
#import "ArgTokenizer.h"
#import "SearchConfig.h"
#import "Regex.h"
#import "SearchOptions.h"

@interface SearchOptions ()
// private properties
@property NSArray<SearchOption*> *searchOptions;
@property NSDictionary<NSString*,NSString*> *longArgDict;
@property NSDictionary *boolActionDict;
@property NSDictionary *stringActionDict;
@property NSDictionary *integerActionDict;
@property ArgTokenizer *argTokenizer;

@end

@implementation SearchOptions

- (instancetype) init {
    self = [super init];
    if (self) {
        self.longArgDict = [self getLongArgDict];
        self.boolActionDict = [self getBoolActionDict];
        self.stringActionDict = [self getStringActionDict];
        self.integerActionDict = [self getIntegerActionDict];
        self.searchOptions = [self searchOptionsFromJson];
        self.argTokenizer = [[ArgTokenizer alloc] initWithOptions:self.searchOptions];
    }
    return self;
}

- (NSArray<SearchOption*>*) searchOptionsFromJson {
    NSMutableString *searchOptionsJsonPath = [NSMutableString stringWithString:getXsearchSharedPath()];
    [searchOptionsJsonPath appendString:@"/searchoptions.json"];
    
    if (![[NSFileManager defaultManager] fileExistsAtPath:searchOptionsJsonPath]) {
        return nil;
    }
    
    NSMutableArray *searchOptions = [[NSMutableArray alloc] initWithCapacity:44];

    NSData *data = [NSData dataWithContentsOfFile:searchOptionsJsonPath];
    
    if (NSClassFromString(@"NSJSONSerialization")) {
        NSError *error = nil;
        id jsonObject = [NSJSONSerialization
                         JSONObjectWithData:data
                         options:0
                         error:&error];
        
        if (error) { /* JSON was malformed, act appropriately here */ }
        
        if ([jsonObject isKindOfClass:[NSDictionary class]]) {
            NSArray *searchOptionObjects = jsonObject[@"searchoptions"];
            for (NSDictionary *searchOptionDict in searchOptionObjects) {
                NSString *lArg = searchOptionDict[@"long"];
                NSString *sArg = searchOptionDict[@"short"];
                NSString *desc = searchOptionDict[@"desc"];
                ArgTokenType argType = ArgTokenTypeUnknown;
                if (self.boolActionDict[lArg]) {
                    argType = ArgTokenTypeBool;
                } else if (self.stringActionDict[lArg] || [lArg isEqualToString:@"settings-file"]) {
                    argType = ArgTokenTypeStr;
                } else if (self.integerActionDict[lArg]) {
                    argType = ArgTokenTypeInt;
                }
                [searchOptions addObject:[[SearchOption alloc] initWithShortArg:sArg withLongArg:lArg withDesc:desc withArgType:argType]];
            }
        }
    }

    return [NSArray arrayWithArray:searchOptions];
}

- (NSDictionary<NSString*,NSString*>*) getLongArgDict {
    NSMutableDictionary *longArgDict = [[NSMutableDictionary alloc] initWithCapacity:70];
    for (SearchOption *so in self.searchOptions) {
        longArgDict[so.longArg] = so.longArg;
        if (so.shortArg) {
            longArgDict[so.shortArg] = so.longArg;
        }
    }
    // Add path here because it isn't included in findoptions.json
    longArgDict[@"path"] = @"path";
    return [NSDictionary dictionaryWithDictionary:longArgDict];
}

typedef void (^BoolActionBlockType)(BOOL, SearchSettings*);

- (NSDictionary<NSString*,BoolActionBlockType>*) getBoolActionDict {
    return @{
        @"allmatches" : [^void (BOOL b, SearchSettings *ss) { ss.firstMatch = !b; } copy],
        @"archivesonly" : [^void (BOOL b, SearchSettings *ss) {
            ss.archivesOnly = b;
            if (b) ss.searchArchives = true;
        } copy],
        @"colorize" : [^void (BOOL b, SearchSettings *ss) { ss.colorize = b; } copy],
        @"debug" : [^void (BOOL b, SearchSettings *ss) {
            ss.debug = b;
            if (b) ss.verbose = true;
        } copy],
        @"excludehidden" : [^void (BOOL b, SearchSettings *ss) { ss.includeHidden = !b; } copy],
        @"firstmatch" : [^void (BOOL b, SearchSettings *ss) { ss.firstMatch = b; } copy],
        @"followsymlinks" : [^void (BOOL b, SearchSettings *ss) { ss.followSymlinks = b; } copy],
        @"help" : [^void (BOOL b, SearchSettings *ss) { ss.printUsage = b; } copy],
        @"includehidden" : [^void (BOOL b, SearchSettings *ss) { ss.includeHidden = b; } copy],
        @"multilinesearch" : [^void (BOOL b, SearchSettings *ss) { ss.multiLineSearch = b; } copy],
        @"nocolorize" : [^void (BOOL b, SearchSettings *ss) { ss.colorize = !b; } copy],
        @"nofollowsymlinks" : [^void (BOOL b, SearchSettings *ss) { ss.followSymlinks = !b; } copy],
        @"noprintdirs" : [^void (BOOL b, SearchSettings *ss) { ss.printDirs = !b; } copy],
        @"noprintfiles" : [^void (BOOL b, SearchSettings *ss) { ss.printFiles = !b; } copy],
        @"noprintlines" : [^void (BOOL b, SearchSettings *ss) { ss.printLines = !b; } copy],
        @"noprintmatches" : [^void (BOOL b, SearchSettings *ss) { ss.printResults = !b; } copy],
        @"norecursive" : [^void (BOOL b, SearchSettings *ss) { ss.recursive = !b; } copy],
        @"nosearcharchives" : [^void (BOOL b, SearchSettings *ss) { ss.searchArchives = !b; } copy],
        @"printdirs" : [^void (BOOL b, SearchSettings *ss) { ss.printDirs = b; } copy],
        @"printfiles" : [^void (BOOL b, SearchSettings *ss) { ss.printFiles = b; } copy],
        @"printlines" : [^void (BOOL b, SearchSettings *ss) { ss.printLines = b; } copy],
        @"printmatches" : [^void (BOOL b, SearchSettings *ss) { ss.printResults = b; } copy],
        @"recursive" : [^void (BOOL b, SearchSettings *ss) { ss.recursive = b; } copy],
        @"searcharchives" : [^void (BOOL b, SearchSettings *ss) { ss.searchArchives = b; } copy],
        @"sort-ascending" : [^void (BOOL b, SearchSettings *ss) { ss.sortDescending = !b; } copy],
        @"sort-caseinsensitive" : [^void (BOOL b, SearchSettings *ss) { ss.sortCaseInsensitive = b; } copy],
        @"sort-casesensitive" : [^void (BOOL b, SearchSettings *ss) { ss.sortCaseInsensitive = !b; } copy],
        @"sort-descending" : [^void (BOOL b, SearchSettings *ss) { ss.sortDescending = b; } copy],
        @"uniquelines" : [^void (BOOL b, SearchSettings *ss) { ss.uniqueLines = b; } copy],
        @"verbose" : [^void (BOOL b, SearchSettings *ss) { ss.verbose = b; } copy],
        @"version" : [^void (BOOL b, SearchSettings *ss) { ss.printVersion = b; } copy]
    };
}

typedef void (^StringActionBlockType)(NSString*, SearchSettings*);

- (NSDictionary<NSString*,StringActionBlockType>*) getStringActionDict {
    return @{
        @"encoding" : ^void (NSString* s, SearchSettings *ss) {
            ss.textFileEncoding = s;
        },
        @"in-archiveext" : ^void (NSString* s, SearchSettings *ss) {
            [ss addInArchiveExtension:s];
        },
        @"in-archivefilepattern" : ^void (NSString* s, SearchSettings *ss) {
            [ss.inArchiveFilePatterns addObject:[[Regex alloc] initWithPattern:s]];
        },
        @"in-dirpattern" : ^void (NSString* s, SearchSettings *ss) {
            [ss.inDirPatterns addObject:[[Regex alloc] initWithPattern:s]];
        },
        @"in-ext" : ^void (NSString* s, SearchSettings *ss) {
            [ss addInExtension:s];
        },
        @"in-filepattern" : ^void (NSString* s, SearchSettings *ss) {
            [ss.inFilePatterns addObject:[[Regex alloc] initWithPattern:s]];
        },
        @"in-filetype" : ^void (NSString* s, SearchSettings *ss) {
            [ss addInFileType:s];
        },
        @"in-linesafterpattern" : ^void (NSString* s, SearchSettings *ss) {
            [ss.inLinesAfterPatterns addObject:[[Regex alloc] initWithPattern:s]];
        },
        @"in-linesbeforepattern" : ^void (NSString* s, SearchSettings *ss) {
            [ss.inLinesBeforePatterns addObject:[[Regex alloc] initWithPattern:s]];
        },
        @"linesaftertopattern" : ^void (NSString* s, SearchSettings *ss) {
            [ss.linesAfterToPatterns addObject:[[Regex alloc] initWithPattern:s]];
        },
        @"linesafteruntilpattern" : ^void (NSString* s, SearchSettings *ss) {
            [ss.linesAfterUntilPatterns addObject:[[Regex alloc] initWithPattern:s]];
        },
        @"maxlastmod" : ^void (NSString* s, SearchSettings *ss) {
            [ss setMaxLastModFromString:s];
        },
        @"minlastmod" : ^void (NSString* s, SearchSettings *ss) {
            [ss setMinLastModFromString:s];
        },
        @"out-archiveext" : ^void (NSString* s, SearchSettings *ss) {
            [ss addOutArchiveExtension:s];
        },
        @"out-archivefilepattern" : ^void (NSString* s, SearchSettings *ss) {
            [ss.outArchiveFilePatterns addObject:[[Regex alloc] initWithPattern:s]];
        },
        @"out-dirpattern" : ^void (NSString* s, SearchSettings *ss) {
            [ss.outDirPatterns addObject:[[Regex alloc] initWithPattern:s]];
        },
        @"out-ext" : ^void (NSString* s, SearchSettings *ss) {
            [ss addOutExtension:s];
        },
        @"out-filepattern" : ^void (NSString* s, SearchSettings *ss) {
            [ss.outFilePatterns addObject:[[Regex alloc] initWithPattern:s]];
        },
        @"out-filetype" : ^void (NSString* s, SearchSettings *ss) {
            [ss addOutFileType:s];
        },
        @"out-linesafterpattern" : ^void (NSString* s, SearchSettings *ss) {
            [ss.outLinesAfterPatterns addObject:[[Regex alloc] initWithPattern:s]];
        },
        @"out-linesbeforepattern" : ^void (NSString* s, SearchSettings *ss) {
            [ss.outLinesBeforePatterns addObject:[[Regex alloc] initWithPattern:s]];
        },
        @"path" : ^void (NSString* s, SearchSettings *ss) {
            [ss addPath:s];
        },
        @"searchpattern" : ^void (NSString* s, SearchSettings *ss) {
            [ss.searchPatterns addObject:[[Regex alloc] initWithPattern:s]];
        },
        @"sort-by" : ^void (NSString* s, SearchSettings *ss) {
            [ss setSortByFromName:s];
        }
    };
}

typedef void (^IntegerActionBlockType)(NSInteger, SearchSettings*);

- (NSDictionary<NSString*,IntegerActionBlockType>*) getIntegerActionDict {
    return @{
        @"linesafter" : [^void (NSInteger i, SearchSettings *ss) { ss.linesAfter = i; } copy],
        @"linesbefore" : [^void (NSInteger i, SearchSettings *ss) { ss.linesBefore = i; } copy],
        @"maxdepth" : [^void (NSInteger i, SearchSettings *ss) { ss.maxDepth = i; } copy],
        @"maxlinelength" : [^void (NSInteger i, SearchSettings *ss) { ss.maxLineLength = i; } copy],
        @"maxsize" : [^void (NSInteger i, SearchSettings *ss) { ss.maxSize = i; } copy],
        @"mindepth" : [^void (NSInteger i, SearchSettings *ss) { ss.minDepth = i; } copy],
        @"minsize" : [^void (NSInteger i, SearchSettings *ss) { ss.minSize = i; } copy]
    };
}

- (void) applyArgTokenToSettings:(ArgToken *)argToken settings:(SearchSettings *)settings error:(NSError **)error {
    if (argToken.type == ArgTokenTypeBool) {
        if ([argToken.value isKindOfClass:[NSNumber class]]) {
            NSNumber *num = (NSNumber *)argToken.value;
            BOOL b = [num boolValue];
            void(^block)(BOOL, FindSettings*) = self.boolActionDict[argToken.name];
            block(b, settings);
        } else {
            setError(error, [@"Invalid value for option: " stringByAppendingString:argToken.name]);
            return;
        }
    } else if (argToken.type == ArgTokenTypeStr) {
        if ([argToken.value isKindOfClass:[NSString class]]) {
            NSString *s = (NSString *)argToken.value;
            void(^block)(NSString *, FindSettings *) = self.stringActionDict[argToken.name];
            block(s, settings);
        } else if ([argToken.value isKindOfClass:[NSArray class]]) {
            NSArray *arr = (NSArray *)argToken.value;
            for (NSObject *o in arr) {
                if ([o isKindOfClass:[NSString class]]) {
                    NSString *s = (NSString *)argToken.value;
                    void(^block)(NSString *, FindSettings *) = self.stringActionDict[argToken.name];
                    block(s, settings);
                } else {
                    setError(error, [@"Invalid value for option: " stringByAppendingString:argToken.name]);
                    return;
                }
            }
        } else {
            setError(error, [@"Invalid value for option: " stringByAppendingString:argToken.name]);
            return;
        }
    } else if (argToken.type == ArgTokenTypeInt) {
        if ([argToken.value isKindOfClass:[NSNumber class]]) {
            NSNumber *num = (NSNumber *)argToken.value;
            NSInteger i = [num integerValue];
            void(^block)(NSInteger, FindSettings*) = self.integerActionDict[argToken.name];
            block(i, settings);
        } else {
            setError(error, [@"Invalid value for option: " stringByAppendingString:argToken.name]);
            return;
        }
    } else {
        setError(error, [@"Invalid option: " stringByAppendingString:argToken.name]);
        return;
    }
}

- (void) updateSettingsFromArgTokens:(SearchSettings *)settings argTokens:(NSArray<ArgToken*> *)argTokens error:(NSError **)error {
    for (ArgToken *argToken in argTokens) {
        if ([argToken.name isEqualToString:@"settings-file"]) {
            [self updateSettingsFromFile:settings filePath:[NSString stringWithFormat:@"%@", argToken.value] error:error];
        } else {
            [self applyArgTokenToSettings:argToken settings:settings error:error];
        }
        if (*error) {
            return;
        }
    }
}

- (void) updateSettingsFromDictionary:(SearchSettings *)settings dictionary:(NSDictionary *)dictionary error:(NSError **)error {
    NSArray<ArgToken*> *argTokens = [self.argTokenizer tokenizeDictionary:dictionary error:error];
    if (*error) {
        return;
    }
    [self updateSettingsFromArgTokens:settings argTokens:argTokens error:error];
}

- (void) updateSettingsFromData:(SearchSettings *)settings data:(NSData *)data error:(NSError **)error {
    NSArray<ArgToken*> *argTokens = [self.argTokenizer tokenizeData:data error:error];
    if (*error) {
        return;
    }
    [self updateSettingsFromArgTokens:settings argTokens:argTokens error:error];
}

- (FindSettings *) settingsFromData:(NSData *)data error:(NSError **)error {
    SearchSettings *settings = [[SearchSettings alloc] init];
    [self updateSettingsFromData:settings data:data error:error];
    return settings;
}

- (void) updateSettingsFromFile:(SearchSettings *)settings filePath:(NSString *)filePath error:(NSError **)error {
    NSArray<ArgToken*> *argTokens = [self.argTokenizer tokenizeFile:filePath error:error];
    if (*error) {
        return;
    }
    [self updateSettingsFromArgTokens:settings argTokens:argTokens error:error];
}

- (SearchSettings *) settingsFromFile:(NSString *)filePath error:(NSError **)error {
    SearchSettings *settings = [[SearchSettings alloc] init];
    [self updateSettingsFromFile:settings filePath:filePath error:error];
    return settings;
}

- (void) updateSettingsFromArgs:(SearchSettings *)settings args:(NSArray *)args error:(NSError **)error {
    NSArray<ArgToken*> *argTokens = [self.argTokenizer tokenizeArgs:args error:error];
    if (*error) {
        return;
    }
    [self updateSettingsFromArgTokens:settings argTokens:argTokens error:error];
}

- (SearchSettings *) settingsFromArgs:(NSArray<NSString*> *)args error:(NSError **)error {
    SearchSettings *settings = [[SearchSettings alloc] init];
    // default printResults to true since running as cli
    settings.printResults = true;
    [self updateSettingsFromArgs:settings args:args error:error];
    return settings;
}

- (NSString*) getUsageString {
    NSMutableString *s = [[NSMutableString alloc] initWithString:@"\nUsage:\n"];
    [s appendString:@" objcsearch [options] -s <searchpattern> <path> [<path> ...]\n\n"];
    [s appendString:@"Options:\n"];
    NSMutableArray *optStrings = [NSMutableArray array];
    long longest = 0;
    NSArray *sortedOptions = [self.searchOptions sortedArrayUsingComparator:^NSComparisonResult(SearchOption *so1, SearchOption *so2) {
        return [[so1 sortArg] compare:[so2 sortArg]];
    }];
    for (SearchOption *so in sortedOptions) {
        NSMutableString *optString = [[NSMutableString alloc] init];
        if (so.shortArg) {
            [optString appendFormat:@"-%@,", so.shortArg];
        }
        [optString appendFormat:@"--%@", so.longArg];
        if ([optString length] > longest) {
            longest = [optString length];
        }
        [optStrings addObject:optString];
    }
    // For some reason, length-specified fields don't work in NSString format strings,
    // so forced to use char * and sprintf
    char *metaString = " %%-%lus  %%s\n";
    char templateString[20];
    sprintf(templateString, metaString, longest);
    for (int i=0; i < [sortedOptions count]; i++) {
        NSString *optString = optStrings[i];
        NSString *optDesc = [sortedOptions[i] desc];
        long formatLen = [optString length] + [optDesc length] + 5;
        char formatString[formatLen];
        sprintf(formatString, templateString, [optString UTF8String], [optDesc UTF8String]);
        [s appendString:[NSString stringWithUTF8String:formatString]];
    }
    return [NSString stringWithString:s];
}

- (void) usage:(int)code {
    logMsg([self getUsageString]);
    exit(code);
}

@end
